Trigger GC when crossing warning threshold on iOS

bnham opened the pull request:
#### ae91c42fd27e0a7f0b05af5edf0245b5b36805d7
<pre>
Trigger GC when crossing warning threshold on iOS
<a href="https://bugs.webkit.org/show_bug.cgi?id=273830">https://bugs.webkit.org/show_bug.cgi?id=273830</a>
<a href="https://rdar.apple.com/127677553">rdar://127677553</a>

Reviewed by Geoffrey Garen.

On iOS where we have fairly strict memory limits, sometimes the mutator can get far ahead of the
GC. We might wait until a PROC_LIMIT_CRITICAL event fires before a full GC occurs, at which point
the process is already eligible to be killed by the OS.

To try and reduce the number of foreground OOMs, let&apos;s also trigger a full GC if it hasn&apos;t occurred
recently when a PROC_LIMIT_WARNING event fires (which is 80% of the critical threshold).

* Source/WebCore/page/MemoryRelease.cpp:
(WebCore::releaseMemory):

Canonical link: <a href="https://commits.webkit.org/278567@main">https://commits.webkit.org/278567@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/4fb18f71091718f2ca91f48160df093a371f6a17

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/50659 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/29956 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/2976 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/53918 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/59/builds/1350 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/52962 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/36216 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/1000 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/41306 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/52758 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/27611 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/43636 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/22424 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/24996 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/892 "Passed tests") | [✅ 🛠 wpe-skia](https://ews-build.webkit.org/#/builders/52/builds/9100 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/46977 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/954 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/55508 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/25761 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/62/builds/857 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/48710 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/27018 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/43779 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/47766 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/11159 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/27886 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/26750 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

geoffreygaren reviewed (approved):
r=me

cdumez commented on Source/WebCore/page/MemoryRelease.cpp:
```diff
@@ -160,6 +160,11 @@ void releaseMemory(Critical critical, Synchronous synchronous, MaintainBackForwa
 {
     TraceScope scope(MemoryPressureHandlerStart, MemoryPressureHandlerEnd, static_cast<uint64_t>(critical), static_cast<uint64_t>(synchronous));
 
+#if PLATFORM(IOS_FAMILY)
```
Shouldn't this go in platformReleaseMemory() in the Cocoa file since this is IOS only?

bnham replied:
Possibly yes. I was trying to put the code in the same file as the other call to `garbageCollectNowIfNotDoneRecently`, which is also scoped to only occurring on IOS_FAMILY.

bnham committed:
Trigger GC when crossing warning threshold on iOS
https://bugs.webkit.org/show_bug.cgi?id=273830
rdar://127677553

Reviewed by Geoffrey Garen.

On iOS where we have fairly strict memory limits, sometimes the mutator can get far ahead of the
GC. We might wait until a PROC_LIMIT_CRITICAL event fires before a full GC occurs, at which point
the process is already eligible to be killed by the OS.

To try and reduce the number of foreground OOMs, let's also trigger a full GC if it hasn't occurred
recently when a PROC_LIMIT_WARNING event fires (which is 80% of the critical threshold).

* Source/WebCore/page/MemoryRelease.cpp:
(WebCore::releaseMemory):

Canonical link: https://commits.webkit.org/278567@main
